src/main/java/io/strimzi/kafka/bridge/mqtt/core/MqttServerHandler.java

```diff
@@ -76,6 +92,14 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
         ctx.close();
     }
 
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        bridgeKafkaProducerWithNoAck.close();
+        bridgeKafkaProducerWithAckOne.close();
+        ctx.close();
+        super.handlerRemoved(ctx);
+    }
```

Comment:
I think this has a problem, not sure how you tested it.
Try to open an MQTT client connection with client 1 and send messages, then DON'T close the connection.
Open another MQTT client connection with client 2 and send messages. Until now everything ok.
Close MQTT connection of client 1, it's going to close the Kafka producers and still connected MQTT client 2 won't be able to send messages anymore.

```shell
java.lang.IllegalStateException: Cannot perform operation after producer has been closed
	at org.apache.kafka.clients.producer.KafkaProducer.throwIfProducerClosed(KafkaProducer.java:969) ~[kafka-clients-3.5.0.jar:?]
	at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:991) ~[kafka-clients-3.5.0.jar:?]
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:962) ~[kafka-clients-3.5.0.jar:?]
	at io.strimzi.kafka.bridge.mqtt.kafka.BridgeKafkaProducer.send(BridgeKafkaProducer.java:45) ~[classes/:?]
	at io.strimzi.kafka.bridge.mqtt.core.MqttServerHandler.handlePublishMessage(MqttServerHandler.java:166) ~[classes/:?]
	at io.strimzi.kafka.bridge.mqtt.core.MqttServerHandler.channelRead0(MqttServerHandler.java:80) ~[classes/:?]
	at io.strimzi.kafka.bridge.mqtt.core.MqttServerHandler.channelRead0(MqttServerHandler.java:39) ~[classes/:?]
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346) ~[netty-codec-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318) ~[netty-codec-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[netty-transport-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

This is because the Kafka producers were closed when MQTT client 1 disconnected.

Reply:
I tested sequentially. First all clients connect, and randomly all of them start producing message to the server, as soon as a client is done producing, it should disconnect. This is kind same the below sequence. Maybe I missed some testing after implementing this callback. Please, allow me some time to investigate further.

Reply:
this callback should not be invoked on a client disconnection, according to the [netty docs ](https://netty.io/4.0/api/io/netty/channel/ChannelHandler.html#:~:text=void-,handlerRemoved(ChannelHandlerContext%C2%A0ctx),removed%20from%20the%20actual%20context%20and%20it%20doesn%27t%20handle%20events%20anymore.,-Method%20Detail), maybe I got it wrong!

Reply:
It is not working as expected :(.

I just connected all clients and without producing messages, I closed all of them, it this is making the producers to get closed. It should not work like! The producers just close in case the handler is destroyed or being closed.

Reply:
I believe there is no a way to do this inside the `SimpleChannelInboundHandler`.
Even when our current handler is sharable among channels, the [callbacks](https://netty.io/4.0/api/io/netty/channel/SimpleChannelInboundHandler.html#:~:text=Nested%20classes/interfaces%20inherited%20from%20interface%C2%A0io.netty.channel.ChannelHandler) have effects with respect to the current channel and not the handler itself. 

To be more clear, a very good [illustration](https://medium.com/geekculture/a-tour-of-netty-5020ecee5494#:~:text=each%20core%20component.-,Netty%20threading%20model%2C%20source%3A%20https%3A//blog.e%2Dkursy.it/netty%2Dintroduction/,-Netty%20ByteBuf) shows how channels are handled among event loops (worker groups). 

![image](https://github.com/strimzi/strimzi-mqtt-bridge/assets/42675180/9b23dca9-279a-4418-a05d-620a08df651e)

We have different pipelines for different channels (open connection), and each of them have a `ChannelHandler`, in our case, the `ChannelHandler` is `sharable`, so the same instance of it are being used among different channels, which means if we close the producer anywhere inside the `ChannelHandler`, we are going to close for all. 

Maybe we should consider providing global state and close the producers on the bridge shutdown?